Libplanet/Blockchain/BlockChain.cs

```diff
@@ -615,21 +635,9 @@ public void StageTransaction(Transaction<T> transaction)
                     msg);
             }
 
-            // FIXME it's global chain lock so using it in this method can cause degrading
-            // parallelism of `BlockChain<T>`. we should re-organize locks in `BlockChain<T>`
-            _rwlock.EnterWriteLock();
-
-            try
-            {
-                if (!_transactions.ContainsKey(transaction.Id))
-                {
-                    _transactions[transaction.Id] = transaction;
-                    Store.StageTransactionIds(ImmutableHashSet.Create(transaction.Id));
-                }
-            }
-            finally
+            if (!StagePolicy.HasStaged(this, transaction.Id, includeUnstaged: true))
             {
-                _rwlock.ExitWriteLock();
+                StagePolicy.Stage(this, transaction);
```

Comment:
Is `StagePolicy.Stage()` thread-safe?

Reply:
It should be so (interface-wise), and I believe `VolatileStagePolicy<T>`, the only implementation, is so.  (It uses `ConcurrentDictionary<K, V>` under the hood.)

Reply:
OK, if `.Unstage()` is too, it seems that we can strip `_rwlock` in `BlockChain<T>.UnstageTransaction()`.